fix: handle versioned cloud URLs in reboot flow

adamshiervani opened the pull request.

adamshiervani committed:
fix: handle versioned cloud URLs in reboot flow

Support redirecting to correct versioned path after OTA updates in cloud mode.
Local mode uses HTTP polling, cloud mode uses WebRTC reconnection + RPC.

adamshiervani committed:
refactor: centralize cloud URL construction logic

Extracted cloud URL building logic into a reusable function, `buildCloudUrl`, to streamline KVM and VideoOverlay components. This improves code maintainability and reduces duplication of version handling for cloud URLs.

IDisposable commented on ui/src/components/VideoOverlay.tsx:
```diff
@@ -398,124 +399,126 @@ export function PointerLockBar({ show }: PointerLockBarProps) {
 
+  // Redirect helper - navigates and forces reload
+  const redirectTo = useCallback(async (url: string) => {
+    window.location.href = url;
```
Glad you extracted this pattern out to a utils.ts function... I think I've see it before ;)

IDisposable replied:
Should we stick in a console.log here?

IDisposable commented on ui/src/components/VideoOverlay.tsx:
```diff
@@ -398,124 +399,126 @@ export function PointerLockBar({ show }: PointerLockBarProps) {
 
-      console.log('Checking post-reboot health endpoint:', postRebootAction.healthCheck);
-      const timeoutId = window.setTimeout(() => abortController.abort(), 2000);
+      const timeout = setTimeout(() => controller.abort(), 2000);
```
Is 2 seconds enough?

IDisposable commented on ui/src/components/VideoOverlay.tsx:
```diff
@@ -398,124 +399,126 @@ export function PointerLockBar({ show }: PointerLockBarProps) {
+
+  // Cloud mode: wait for WebRTC reconnection via RPC, then redirect with versioned URL
+  useEffect(() => {
+    if (isOnDevice || !postRebootAction || !deviceId || !show || !hasSeenDisconnect) return;
```
I think this is muddled enough to warrant two statements:
```typescript
// cloud mode only
if (isOnDevice) return;

// ensure we actually should be waiting
if (!postRebootAction || !deviceId || !show || !hasSeenDisconnect) return;
```

IDisposable replied:
... or even the second line could be 
```typescript
if (false === (postRebootAction && deviceId && show && hasSeenDisconnect)) return;
```

IDisposable commented on ui/src/components/VideoOverlay.tsx:
```diff
@@ -398,124 +399,126 @@ export function PointerLockBar({ show }: PointerLockBarProps) {
+      try {
+        // getLocalVersion internally waits for RTC channel to be ready
+        const { appVersion } = await getLocalVersion();
+        if (cancelled) return;
```
_note to self_: do this second so that during the `await` if it gets cancelled...

IDisposable commented on ui/src/routes/devices.$id.tsx:
```diff
@@ -699,15 +699,13 @@ export default function KvmIdRoute() {
-          healthCheck: postRebootAction?.healthCheck || `${window.location.origin}/device/status`,
-          redirectTo: postRebootAction?.redirectTo || window.location.href,
-        }
+          healthCheck: action?.healthCheck ?? "/device/status",
```
... the this simplifies to 
```typescript
postRebootAction: action || defaultAction,
```

IDisposable commented on ui/src/routes/devices.$id.tsx:
```diff
@@ -699,15 +699,13 @@ export default function KvmIdRoute() {
-          redirectTo: postRebootAction?.redirectTo || window.location.href,
-        }
+          healthCheck: action?.healthCheck ?? "/device/status",
+          redirectTo: action?.redirectTo ?? "/",
```
... cautionary note that if we get an `action` without one of the properties, this could be wrong... but I would contend that the handler that uses the `postRebootAction` should already be hardened against that...

IDisposable commented on ui/src/routes/devices.$id.tsx:
```diff
@@ -823,7 +821,7 @@ export default function KvmIdRoute() {
     // Rebooting takes priority over connection status
     if (rebootState?.isRebooting) {
-      return <RebootingOverlay show={true} postRebootAction={rebootState.postRebootAction} />;
+      return <RebootingOverlay show={true} postRebootAction={rebootState.postRebootAction} deviceId={params.id} />;
```
_unrelated nit_ I would love to see the various "overlays" all have  close/dismiss in case they're wrong/intrusive to see what's going on in the video...

adamshiervani committed:
refactor: enhance reboot flow and local version retrieval

Updated the rebooting overlay to improve the handling of versioned URLs and added a timeout option for local version retrieval. Simplified condition checks in the useEffect hook and added logging for redirection actions. This enhances the clarity and maintainability of the code.

adamshiervani committed:
refactor: fix type issues with getLocalVersionRPC()
